File: src/main/java/voidchess/figures/Queen.java
package voidchess.figures;

import voidchess.board.BasicChessGameInterface;
import voidchess.board.SimpleChessBoardInterface;
import voidchess.helper.Move;
import voidchess.helper.Position;
import voidchess.image.ImageType;

import java.util.List;

/**
 * @author stephan
 */
public class Queen extends Figure {
    private Bishop bishop;
    private Rock rock;

    public Queen(boolean isWhite, Position startPosition) {
        super(isWhite, startPosition, FigureType.QUEEN);
        bishop = new Bishop(isWhite, startPosition);
        rock = new Rock(isWhite, startPosition);
    }

    public void figureMoved(Move move) {
        bishop.figureMoved(move);
        rock.figureMoved(move);
        super.figureMoved(move);
    }

    public void undoMove(Position oldPosition) {
        bishop.undoMove(oldPosition);
        rock.undoMove(oldPosition);
        super.undoMove(oldPosition);
    }

    public boolean isReachable(Position to, BasicChessGameInterface game) {
        return bishop.isReachable(to, game) || rock.isReachable(to, game);
    }

    public void getReachableMoves(BasicChessGameInterface game, List<Move> result) {
        bishop.getReachableMoves(game, result);
        rock.getReachableMoves(game, result);
    }

    public boolean isSelectable(SimpleChessBoardInterface game) {
        return rock.isSelectable(game) || bishop.isSelectable(game);
    }

    public int countReachableMoves(BasicChessGameInterface game) {
        return rock.countReachableMoves(game) + bishop.countReachableMoves(game);
    }
}

File: src/main/kotlin/voidchess/figures/Queen.kt
package voidchess.figures

import voidchess.board.BasicChessBoard
import voidchess.board.ChessBoard
import voidchess.board.check.BoundLine
import voidchess.board.check.CheckLine
import voidchess.board.move.Direction
import voidchess.board.move.Move
import voidchess.board.move.Position


class Queen(isWhite: Boolean, startPosition: Position) : Figure(isWhite, startPosition, FigureType.QUEEN, true, true) {

    override fun isReachable(toPos: Position, game: BasicChessBoard): Boolean {
        val direction = position.getDirectionTo(toPos) ?: return false

        forEachReachablePos(game, direction) {
            if (it.equalsPosition(toPos)) return true
        }

        return false
    }

    private inline fun forEachReachablePos(game: BasicChessBoard, informOf: (Position) -> Unit) {
        forEachReachablePos(game, Direction.UP, informOf)
        forEachReachablePos(game, Direction.LEFT, informOf)
        forEachReachablePos(game, Direction.DOWN, informOf)
        forEachReachablePos(game, Direction.RIGHT, informOf)
        forEachReachablePos(game, Direction.UP_RIGHT, informOf)
        forEachReachablePos(game, Direction.UP_LEFT, informOf)
        forEachReachablePos(game, Direction.DOWN_RIGHT, informOf)
        forEachReachablePos(game, Direction.DOWN_LEFT, informOf)
    }

    override fun getReachableMoves(game: BasicChessBoard, result: MutableList<Move>) {
        forEachReachablePos(game) {
            result.add(Move[position, it])
        }
    }

    override fun getPossibleMovesWhileUnboundAndCheck(game: ChessBoard, checkLine: CheckLine, result: MutableList<Move>) {
        // a queen can intersect a check at up to three positions. but the overhead to check this upper bound is probably to high
        checkLine.posProgression.forEachReachablePos {checkInterceptPos->
            addMoveIfReachable(checkInterceptPos, game, result)
        }
    }

    override fun getPossibleMovesWhileBoundAndNoCheck(game: ChessBoard, boundLine: BoundLine, result: MutableList<Move>) {
        boundLine.possibleMovesToAttacker.forEachReachablePos {posBetweenThisAndAttacker->
            result.add(Move[position, posBetweenThisAndAttacker])
        }
        boundLine.possibleMovesToKing.forEachReachablePos {posBetweenThisAndKing->
            result.add(Move[position, posBetweenThisAndKing])
        }
    }

    override fun isSelectable(game: ChessBoard): Boolean {
        forEachReachablePos(game) {
            if (!isBound(it, game)) return true
        }
        return false
    }

    override fun countReachableMoves(game: BasicChessBoard): Int {
        var reachableMovesCount = 0
        forEachReachablePos(game) {
            reachableMovesCount++
        }
        return reachableMovesCount
    }
}
